crates/typst/src/engine.rs

```diff
@@ -146,20 +186,90 @@ impl Sink {
         Self::default()
     }
 
+    /// Extend the destination sink with the data from the source sink.
+    /// This calls a tracked function on the destination unless the source
+    /// is fully empty (which is usually the case).
+    pub fn extend_tracked(destination: &mut TrackedMut<'_, Self>, source: Sink) {
+        let Sink { delayed, warnings, values, .. } = source;
+        if !delayed.is_empty() || !warnings.is_empty() || !values.is_empty() {
+            destination.extend(delayed, warnings, values);
+        }
+    }
+
     /// Get the stored delayed errors.
     pub fn delayed(&mut self) -> EcoVec<SourceDiagnostic> {
         std::mem::take(&mut self.delayed)
     }
 
-    /// Get the stored warnings.
-    pub fn warnings(self) -> EcoVec<SourceDiagnostic> {
-        self.warnings
-    }
-
     /// Get the values for the traced span.
     pub fn values(self) -> EcoVec<(Value, Option<Styles>)> {
         self.values
     }
+
+    /// Apply warning suppression, deduplication, and return the remaining
+    /// warnings.
+    ///
+    /// We deduplicate warnings which are identical modulo tracepoints.
+    /// This is so we can attempt to suppress each tracepoint separately,
+    /// without having one suppression discard all other attempts of raising
+    /// this same warning through a different set of tracepoints.
+    ///
+    /// For example, calling the same function twice but suppressing a warning
+    /// on the first call shouldn't suppress on the second, so each set of
+    /// tracepoints for a particular warning matters. If at least one instance
+    /// of a warning isn't suppressed, the warning will be returned, and the
+    /// remaining duplicates are discarded.
+    pub fn suppress_and_deduplicate_warnings(
+        mut self,
+        world: &dyn World,
+    ) -> EcoVec<SourceDiagnostic> {
+        let mut unsuppressed_warning_set = HashSet::<u128>::default();
```

Comment:
I think variable names could be less verbose sometimes. E.g. `unsuppressed` would be sufficient imo. At least the `_set` suffix is unnecessary since we can see the type.